kernel/spawn/src/lib.rs

```diff
@@ -982,31 +987,29 @@ fn remove_current_task_from_runqueue(current_task: &ExitableTaskRef) {
         }
         // The regular spill-free version does brute-force removal of the task from ALL runqueues.
         #[cfg(not(runqueue_spillful))] {
-            runqueue::remove_task_from_all(current_task).unwrap();
+            scheduler::remove_task_from_all(current_task).unwrap();
         }
     }
 
     // In the regular case, we do not perform task migration between cores,
     // so we can use the heuristic that the task is only on the current core's runqueue.
     #[cfg(not(rq_eval))] {
-        if let Err(e) = runqueue::get_runqueue(cpu::current_cpu())
-            .ok_or("couldn't get this core's ID or runqueue to remove exited task from it")
-            .and_then(|rq| rq.write().remove_task(current_task)) 
-        {
-            error!("BUG: couldn't remove exited task from runqueue: {}", e);
+        if let Some(run_queue) = scheduler::get_run_queue(cpu::current_cpu()) {
+            run_queue.write().remove(current_task);
+        } else {
+            error!("BUG: couldn't get run queue to remove exited task");
         }
     }
 }
 
-/// Spawns an idle task on the current CPU and adds it to this CPU's runqueue.
-pub fn create_idle_task() -> Result<JoinableTaskRef, &'static str> {
-    let apic_id = cpu::current_cpu();
+pub fn create_idle_task(apic_id: u8) -> Result<TaskRef, &'static str> {
     debug!("Spawning a new idle task on core {}", apic_id);
 
     new_task_builder(idle_task_entry, apic_id)
         .name(format!("idle_task_core_{}", apic_id))
         .idle(apic_id)
         .spawn_restartable(None)
+        .map(|task| task.clone())
```

Comment:
this is correct, but quite tricky to understand. Most readers would likely misinterpret it as a "dumb unnecessary clone". 

It would be better to clearly show what is happening (or at least add a comment) explaining that we are dropping the `JoinableTaskRef` on purpose and obtaining a `TaskRef` since we do not want to join in later.

Even better would be to introduce a function like
```rust
impl JoinableTaskRef {
    /// Drops this `JoinableTaskRef`, making the task an orphan by ensuring
    /// that it can never be joined.
    ///
    /// This guarantees that it will be auto-reaped upon exit.
    pub fn orphan(self) -> TaskRef {
        ...
    }
}
```
and then call that `orphan()` function here.

Reply:
perhaps `into_orphan()` would be more Rusty

Reply:
Added in `d6324142`. I think `orphan` is better because it's not just converting the `JoinableTaskRef` but also modifying the internal state.